docs/RFCS/distributed_csv.md

```diff
@@ -0,0 +1,126 @@
+- Feature Name: Distributed CSV Loading
+- Status: draft
+- Start Date: 2017-07-11
+- Authors: Matt Jibson, Daniel Harrison
+- RFC PR:
+- Cockroach Issue:
+
+# Summary
+
+Add an `LOAD CSV` SQL statement that reads CSV files and converts them into enterprise BACKUP format. These can then be restored via the normal RESTORE statement.
+
+# Motivation
+
+Users desiring to try CockroachDB need a fast way to import their data. CSV appears to be the most common format between databases and it is the first we have decided to support.
+
+As a first implementation, we are assuming that CSV files exist not in cloud storage but instead evenly on each node. For example, if there is some existing Cassandra cluster, each node exports whatever data it is in charge of into CSV on its local disk. We also assume there is enough spare disk space for the processing and importing of this data. A later feature will allow for CSV files to be in cloud storage. Some pre-processor will have to split up these CSV files into chunks (since it could also be just a single, large CSV file) and divide up those chunks among nodes. This would mean adding a seek or read-at-position method to our export store interface. However after this step, the remainder is the same.
+
+# Detailed design
+
+The backup file format is set of SST files. These files must be non-overlapping. The input CSV files may be in some order, but we need to put them into CockroachDB-sorted order so that we can create non-overlapping files. The difficulty then is sorting a large number and size of CSV files (where size is larger than the memory on a single or even all of the machines).
+
+A coordinator node receives a SQL command to import a set of CSVs that represent a single table from a specific directory that is relative to each machine, for example:
+
+`LOAD CSV FROM "/data/csv"`
+
+The coordinator instructs (with some new RPCs) all nodes to read all CSV files from the `/data/csv` directory local to each of them. The nodes read all of their CSV files, convert each row into KVs, and write those KVs to a single RocksDB instance. The cockroach store temp directory is used as temp space for the RocksDB instance.
```

Comment:
As discussed in person, it is probably better to read the CSV to get a sample of the keys and then read it again to distribute, rather than to read the CSV, insert into RocksDB and then read from RocksDB. That's a guess, though, and something that can be determined during implementation.

As a general comment on the RFC, I think this section should be composed of a series of numbered high-level with details about the implementation. For example:

1. Read CSV files to sample keys.
1. Send key sample to coordinator.
1. Coordinator determines split points and initiates global sorting.
1. Read CSV files and convert to KV format and distribute according to global sorting.